### PR TITLE
[Feature] Support data encrypt/decrypt

### LINKED ISSUE #5954: [Proposal] support data encrypt/decrypt
# Data encrypt/decrypt

Many users' sensitive data needs to be stored with data encryption to prevent eavesdropping and interception during transmission, such as banks.

# Design

1. Supports common encryption and decryption and message digest algorithms
2. Key cannot appear in the function plaintext

Implement the syntax for creating a key or public key, and use the `key key_name` instead of the key.

# Detailed Design

## Syntax

The main changes involved here include.

1. implement the class for create/show/drop keys
2. persist the key to the database, involving catalog-related changes
3. parse the key_name as a string when using

```sql
-- create
CREATE ENCRYPTKEY my_key as "ABCD123456789";

-- show 
SHOW ENCRYPTKEYS;

-- drop 
DROP ENCRYPTKEY my_key;

-- usage
SELECT HEX(AES_ENCRYPT("Doris is Great", KEY my_key));

SELECT AES_DECRYPT(UNHEX('D26DB38579D6A343350EDDC6F2AD47C6'), KEY my_key);

### PR DESCRIPTION
**Pull request by qidaye**
## Proposed changes

Add support for data encryption/decryption.

## Types of changes

What types of changes does your code introduce to Doris?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Documentation Update (if none of the other choices apply)
- [ ] Code refactor (Modify the code structure, format the code, etc...)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have created an issue on (Fix #5954) and described the bug/feature there in detail
- [x] Compiling and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If these changes need document changes, I have updated the document
- [ ] Any dependent changes have been merged

## Further comments

If this is a relatively large or complex change, kick off the discussion at dev@doris.apache.org by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Database.java [hunk 1]
```diff
@@ -512,6 +515,13 @@ public void write(DataOutput out) throws IOException {
             }
         }
 
+        // write encryptKeys
+        out.writeInt(name2EncryptKey.size());
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Database.java:519]: create a new class to contains this `name2EncryptKey` for better extension in future.
And using GSON to serde this new class.

File: fe/fe-core/src/main/java/org/apache/doris/catalog/EncryptKey.java [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    public EncryptKeyName getEncryptKeyName() {
+        return encryptKeyName;
+    }
+
+    public String getKeyString() {
+        return keyString;
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/EncryptKey.java:50]: Using GSON serde

File: fe/fe-core/src/main/java/org/apache/doris/catalog/EncryptKeySearchDesc.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+    @Override
+    public String toString() {
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append(encryptKeyName.toString());
+        return stringBuilder.toString();
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/EncryptKeySearchDesc.java:56]: Use GSON serde

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        if (Strings.isNullOrEmpty(dbName)) {
+            dbName = analyzer.getDefaultDb();
+            if (Strings.isNullOrEmpty(dbName)) {
+                ErrorReport.reportAnalysisException(ErrorCode.ERR_NO_DB_ERROR);
+            }
+        } else {
+            dbName = ClusterNamespace.getFullName(getClusterName(), dbName);
+        }
+
+        if (!Catalog.getCurrentCatalog().getAuth().checkDbPriv(ConnectContext.get(), dbName, PrivPredicate.ADMIN)) {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java:69]: when `dbName` is not set, the check is always false.  So we should analyze `dbName` first.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+            ErrorReport.reportAnalysisException(
+                    ErrorCode.ERR_DB_ACCESS_DENIED, ConnectContext.get().getQualifiedUser(), dbName);
+        }
+    }
+
+    public boolean like(String str) {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java:76]: This is consistent with `ShowFunctionStmt` and does not use the complex function of `ShowLoadStmt`.  Is it necessary to make it so complicated?

File: docs/en/sql-reference/sql-statements/Data Definition/CREATE ENCRYPTKEY.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+If the database name is included in `key_name`, then this custom key will be created in the corresponding database, otherwise this function will be created in the database where the current session is located. The name of the new key cannot be the same as the key that already exists in the corresponding database, otherwise the creation will fail.
+
+## Example
+
+1. Create a custom key
+
+    ```
+    CREATE ENCRYPTKEY my_key as "ABCD123456789";
+	```
```
[the remark this answers is not in the document]
Comment (qidaye) [reply docs/en/sql-reference/sql-statements/Data Definition/CREATE ENCRYPTKEY.md]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/Analyzer.java [hunk 7]
```diff
@@ -1806,4 +1808,23 @@ public void markRefdSlots(Analyzer analyzer, PlanNode planRoot,
...
+        List<EncryptKey> retEncryptKeys = new ArrayList<>();
+        if (Strings.isNullOrEmpty(dbName)) {
+            dbName = getDefaultDb();
+        }
+        if ("".equals(dbName)) {
+            ErrorReport.reportAnalysisException(ErrorCode.ERR_NO_DB_ERROR);
+        } else {
+            dbName = ClusterNamespace.getFullName(getClusterName(), dbName);
+            ConcurrentHashMap<String, Database> fullName2Db = getCatalog().getFullNameToDb();
+            Database database = fullName2Db.get(dbName);
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/Analyzer.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateEncryptKeyStmt.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+// under the License.
+
+package org.apache.doris.analysis;
+
+import org.apache.doris.catalog.EncryptKey;
+import org.apache.doris.common.AnalysisException;
+import org.apache.doris.common.UserException;
+
+/**
+ * create a encryptKey
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CreateEncryptKeyStmt.java:32]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CreateEncryptKeyStmt.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        return encryptKey;
+    }
+
+    @Override
+    public void analyze(Analyzer analyzer) throws AnalysisException, UserException {
+        super.analyze(analyzer);
+
+        encryptKeyName.analyze(analyzer);
+
+        encryptKey = new EncryptKey(encryptKeyName, keyString);
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CreateEncryptKeyStmt.java:78]: done, add check not allow empty or null.

File: fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+        }
+    }
+
+    public EncryptKeyName(String keyName) {
+        this.db = null;
+        this.keyName = keyName.toLowerCase();
+    }
+
+    // used to analyze db element in function name, add cluster
+    public String analyzeDb(Analyzer analyzer) throws AnalysisException {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+            throw new AnalysisException("EncryptKey name can not be empty.");
+        }
+        for (int i = 0; i < keyName.length(); ++i) {
+            if (!isValidCharacter(keyName.charAt(i))) {
+                throw new AnalysisException(
+                        "EncryptKey names must be all alphanumeric or underscore. " +
+                                "Invalid name: " + keyName);
+            }
+        }
+        if (Character.isDigit(keyName.charAt(0))) {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+
+    private boolean isValidCharacter(char c) {
+        return Character.isLetterOrDigit(c) || c == '_';
+    }
+
+    public String getDb() {
+        return db;
+    }
+
+    public void setDb(String db) {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java]: Done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+    @Override
+    public String toString() {
+        if (db == null) {
+            return keyName;
+        }
+        return db + "." + keyName;
+    }
+
+    @Override
+    public void write(DataOutput out) throws IOException {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyName.java:92]: Done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyRef.java [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    public EncryptKeyRef(EncryptKeyName encryptKeyName) {
+        super();
+        this.encryptKeyName = encryptKeyName;
+    }
+
+    public LiteralExpr analyzeEncryptKey(Analyzer analyzer) throws AnalysisException {
+        List<EncryptKey> encryptKeys = analyzer.getEncryptKeysFromDb(encryptKeyName.getDb());
+        for (EncryptKey encryptKey : encryptKeys) {
```
[the remark this answers is not in the document]
Comment (qidaye) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/EncryptKeyRef.java]: done

File: fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (morningman) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/ShowEncryptKeysStmt.java:69]: ok

File: fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java [hunk 15]
```diff
@@ -671,6 +674,10 @@ public static AuditEventProcessor getCurrentAuditEventProcessor() {
         return getCurrentCatalog().getAuditEventProcessor();
     }
 
+    public EncryptKeyManager getEncryptKeyManager() {
+        return encryptKeyManager;
```
Comment (morningman) [inline fe/fe-core/src/main/java/org/apache/doris/catalog/Catalog.java:678]: This also need to be persisted in `saveImage()` of Catalog, and read it in `loadImage()`

**Review by morningman — DISMISSED** [review]
LGTM

**Review by morningman — APPROVED** [review]
LGTM

**Pull request merged**